### PR TITLE
[BIP-98 + BIP-116] MERKLEBRANCHVERIFY

### PR DESCRIPTION
**Pull request by kallewoof**
Implements BIPs [98](https://github.com/bitcoin/bips/blob/master/bip-0098.mediawiki) and [116](https://github.com/bitcoin/bips/blob/master/bip-0116.mediawiki) as policy rules for transaction relay, causing transactions which use NOP4 for things other than Merkle branch verification to be treated as non-standard. This is not the soft-fork activation logic needed to use MERKLEBRANCHVERIFY in production.

MERKLEBRANCHVERIFY is a soft-fork upgradeable opcode that allows script writers to commit to a set of data elements and have one or more of these elements be provided at redemption without having to reveal the entire set. As these data elements can be used to encode policy, such as public keys or validation subscripts, the MERKLEBRANCHVERIFY opcode can be used to overcome size limitation of existing bitcoin script, and combines with tail-call semantics to provide a minimal implementation of the Merkelized Abstract Syntax Tree concept.

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Are there demos of scripts based on this/BIP117 anywhere? eg has someone sat down and rewritten the Lightning scripts or TumbleBit scripts to use this stuff to compare the cost/privacy tradeoffs? What about example usage for hiding multisig?

**Comment by maaku** [comment]
There are some examples inside the BIPs:

https://github.com/bitcoin/bips/blob/master/bip-0116.mediawiki
https://github.com/bitcoin/bips/blob/master/bip-0117.mediawiki

It is specifically anticipated that ALL instances of MAST would be of the following:

> redeemScript: `[TOALTSTACK]*N OVER HASH256 <root> 1 MERKLEBRANCHVERIFY 2DROP 2DROP`
> witness: `<policyScript> <proof> <arg1> ... <argN>`

**Comment by TheBlueMatt** [comment]
Hmm, I was hoping for somewhat larger scripts. Specifically, do you have a list of scripts which are cheaper to do using a merkle-hidden tail call than normally outside of large-multisig examples?

**Comment by maaku** [comment]
Basically every script where you are eliding, at minimum, a hash or pubkey push. Sorry for the obtuseness, but it seems like that describes just about anything and it didn't seem appropriate to overload the BIP with examples of everything beyond what was necessary to illustrate the concept. I would certainly like a lightning developer to show how those scripts could be reformulated to use tail-call semantics, but I'm not up to speed on the latest lightning standards.

**Comment by TheBlueMatt** [comment]
Hmm, well my primary motivation for asking was to get a better sense of the cost of using MBV+tail-call in practice. Comparing the cost of LN + TumbleBit scripts seems like an obvious route to figure that out, as it seems to me the only obvious user of something like this would be small-n large-m multisig, which seems somewhat of a waste.

**Comment by maaku** [comment]
It seems rather to me that the obvious user of something like this is 'everybody' for the fungibility improvement.

**Comment by TheBlueMatt** [comment]
Hence my concern - ideally "everyone" would use something like this for fungibility reasons, but I'm highly skeptical anyone will pay a material difference in fees to do so. Obviously I'm not suggesting we go through all the fun of SegWit again to adjust the discount to make a MAST extra branch free, but getting a good idea of exactly what that cost is for common protocols would be useful.

**Comment by maaku** [comment]
I think your intuition for the cost savings is off. Right now even a 2-of-3 multisig ends up costing about the same as a MAST of 2-of-2 vs having an extra (unused) pubkey. Likewise the 2of2 -or- CSV 1of1. And when we switch to Schnorr signatures basically any multisig becomes cheaper in the MAST construct because they all reduce to a MAST of 1of1's.

**Comment by Beyonderx** [comment]
Concept ACK

**Comment by chiguireitor** [comment]
Conceptual question here: am i reading the code right or is the branch size limited to 9997 branches? Imho, having arbitrarily long branching trees could allow for really interesting stuff

**Comment by kallewoof** [comment]
@chiguireitor I believe the 9997 cap is basically representing the maximum depth in the tree. I think that shrinks if you want to pick multiple entries, though. I could be wrong. @maaku?

**Comment by maaku** [comment]
I had to grep the diff to figure out what @chiguireitor was referring to. I assume it is the comment about MAX_STACK_SIZE limiting the number of elements that can be proven from a Merkle tree at once. That number is actually an error, as it should be 997, not 9997 as MAX_STACK_SIZE is a constant limit of 1,000 elements and 3 of those elements are the count parameter, the root hash, and the proof. It's worth noting that a separate limit, the 520 byte push limitation, prevents pushing a proof for a fully expanded Merkle tree of more than about 1380 elements anyway, so even if the MAX_STACK_SIZE limit were avoided you'd hit another one soon enough.

However I think there is some confusion on @chiguireitor's part. Why would you want to prove 1,000+ items from a tree at once? This is the maximum number of elements that you can prove at one time, NOT the maximum number of elements that can be in a tree, for which there is no practical limit. The 520 byte push limitation prevents you from proving a hash more than 16 levels deep in a tree, but MERKLEBRANCHVERIFY calls can be chained allowing for verification of extremely deep trees, much larger than could be practically constructed. A single chained validation allows for trees consisting of billions of elements, that would take hours to construct on a reasonably powerful CPU.

**Comment by chiguireitor** [comment]
Yeah, my bad. Makes sense now, it's just limiting the "next" level of the tree, so in fact, it's not limiting the overall tree structure at all, excellent.

**Comment by martin-lizner** [comment]
Code will go to 0.16 and softfork signalling will start 0.16.1?

**Comment by kallewoof** [comment]
@martin-lizner There is no consensus yet on whether this code will be merged or not. It's still being debated (e.g. on the mailing list and such).

File: src/crypto/sha256.cpp [hunk 1]
```diff
@@ -231,6 +243,11 @@ void CSHA256::Finalize(unsigned char hash[OUTPUT_SIZE])
     WriteBE64(sizedesc, bytes << 3);
     Write(pad, 1 + ((119 - (bytes % 64)) % 64));
     Write(sizedesc, 8);
+    Midstate(hash, NULL, NULL);
+}
+
+void CSHA256::Midstate(unsigned char hash[OUTPUT_SIZE], uint64_t* len, unsigned char* buffer)
```
Comment (DesWurstes) [inline src/crypto/sha256.cpp]: > `uint64_t* len`

It's `uint64_t* bytes` in the sha256.h

~ thanks to @gubatron

File: src/script/interpreter.cpp [hunk 2]
```diff
@@ -1022,6 +1025,142 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
...
+                            }
+                            branch.m_verify.emplace_back(uint256(vchLeaf));
+                        } else {
+                            branch.m_verify.emplace_back();
+                            CHash256().Write(vchLeaf.data(), vchLeaf.size()).Finalize(branch.m_verify.back().begin());
+                        }
+                    }
+
+                    // Compute Merkle root hash
+                    uint256 result = branch.GetHash();
```
Comment (jl2012) [inline src/script/interpreter.cpp]: should error be returned here for invalid MerkleTree structure?

Comment (kallewoof) [reply src/script/interpreter.cpp]: @maaku Is there a reason why you presumed `!invalid` on this one? If not I think we should check validity and return error on `invalid=true`.

File: src/consensus/merkle.h [hunk 3]
```diff
@@ -35,4 +93,1361 @@ uint256 BlockWitnessMerkleRoot(const CBlock& block, bool* mutated = nullptr);
...
+    /* Write a 3-bit code value */
+    MerkleNodeReference& SetCode(MerkleNode::code_type code);
+
+    /* Read and write the MerkleLink values individually. */
+    inline MerkleLink GetLeft() const
+      { return MerkleNode::m_left_from_code[GetCode()]; }
+    inline MerkleNodeReference& SetLeft(MerkleLink left)
+      { return SetCode(MerkleNode::_get_code(left, GetRight())); }
+
+    MerkleLink GetRight() const
```
Comment (jl2012) [inline src/consensus/merkle.h]: inline?

**Comment by kallewoof** [comment]
@jl2012 Thanks for feedback! I inlined `GetRight`/`SetRight` in d99522a.

File: src/consensus/merkle.cpp [hunk 4]
```diff
@@ -185,3 +263,296 @@ std::vector<uint256> BlockMerkleBranch(const CBlock& block, uint32_t position)
...
+    // opcode, which might potentially reduce the number of code paths
+    // in some scripts.
+    if (m_proof.m_path.empty() && m_verify.empty() && m_proof.m_skip.empty()) {
+        if (invalid) {
+            *invalid = false;
+        }
+        return CHashWriter(SER_GETHASH, PROTOCOL_VERSION).GetHash();
+    }
+
+    // Except for the special case of a 0-node, 0-verify, 0-skip tree,
```
Comment (maaku) [inline src/consensus/merkle.cpp]: I think this is the reason for allowing invalid=true in MBV. Maybe it should be fixed by setting *invalid to the logical OR of the various sizes referenced.

Comment (kallewoof) [reply src/consensus/merkle.cpp]: All tests passed, though. I'm pretty sure you are testing for that case, no?

**Comment by kallewoof** [comment]
The failure is due to a recent rewrite causing `READWRITE(REF(CFlatData(m_skip)));` to fail in `merkle.h`. 

Discussing right now whether it makes a difference to switch to varint over compactsize explicitly as is done right now (I argue it does not make a difference, as the size is only a small fraction of the content).

**Comment by kallewoof** [comment]
I'm closing this for now. If a better suited champion appears I'll help out where I can. Ping @btcdrak @maaku.

**Pull request closed without merging**